File: BeeCrowd/1141-Growing-Strings/solution.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cstring>
#include <algorithm>

using namespace std;

// Constraints
const int MAX_NODES = 1000005; // 10^6 + 5
const int ALPHABET = 26;

// Global Automaton Structures
// We use static arrays for performance (better cache locality, no allocation overhead)
int trieNodes[MAX_NODES][ALPHABET];
int failLink[MAX_NODES];
int dp[MAX_NODES]; // Previously nSubWords
bool isWord[MAX_NODES];
int nodesCount = 1;

// Reset only the row we are about to use (Lazy Clearing)
void clearNode(int node)
{
  memset(trieNodes[node], 0, sizeof(int) * ALPHABET);
  isWord[node] = false;
  dp[node] = 0;
  failLink[node] = 0;
}

void insertWord(const string &s)
{
  int curr = 0; // Root is 0
  for (char c : s)
  {
    int idx = c - 'a';
    if (!trieNodes[curr][idx])
    {
      trieNodes[curr][idx] = nodesCount;
      clearNode(nodesCount); // Clear memory only when allocating
      nodesCount++;
    }
    curr = trieNodes[curr][idx];
  }
  isWord[curr] = true;
}

int buildAutomaton()
{
  queue<int> q;
  int maxChain = 0;

  // 1. Initialize Root's children (Depth 1)
  for (int i = 0; i < ALPHABET; i++)
  {
    int child = trieNodes[0][i];
    if (child)
    {
      // Fail link of depth 1 is always root (0)
      // DP: if it's a word, count starts at 1, else 0.
      if (isWord[child])
      {
        dp[child] = 1;
        maxChain = max(maxChain, 1);
      }
      q.push(child);
    }
  }

  // 2. BFS to build Fail Links and compute DP
  while (!q.empty())
  {
    int u = q.front();
    q.pop();

    for (int i = 0; i < ALPHABET; i++)
    {
      int v = trieNodes[u][i];

      if (v)
      {
        // Case A: Child exists.
        // 1. Set Failure Link: Use the pre-calculated edge from the failure of parent
        failLink[v] = trieNodes[failLink[u]][i];

        // 2. DP Logic:
        // Option 1: Extend from Parent (Prefix property)
        int fromParent = dp[u];
        // Option 2: Extend from Failure Link (Suffix property)
        int fromSuffix = dp[failLink[v]];

        dp[v] = max(fromParent, fromSuffix);

        if (isWord[v])
        {
          dp[v]++;
        }

        maxChain = max(maxChain, dp[v]);
        q.push(v);
      }
      else
      {
        // Case B: Child doesn't exist.
        // Optimization: Direct the edge to the failure state's edge.
        // This creates the DFA Graph Structure (O(1) transition).
        trieNodes[u][i] = trieNodes[failLink[u]][i];
      }
    }
  }
  return maxChain;
}

int main()
{
  // Fast I/O
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  while (cin >> n && n != 0)
  {
    // Reset for new test case
    nodesCount = 1;
    clearNode(0); // Clear root manually

    string s;
    for (int i = 0; i < n; i++)
    {
      cin >> s;
      insertWord(s);
    }

    cout << buildAutomaton() << "\n";
  }
  return 0;
}

File: BeeCrowd/1141-Growing-Strings/solution.js
"use strict";
const fs = require("fs");

// -------------------------------------------------------------------------
// CONSTANTS & MEMORY ALLOCATION
// -------------------------------------------------------------------------
// Max characters sum is 10^6. Trie nodes <= 10^6 + 1.
const MAX_NODES = 1000005;
const ALPHABET = 26;

// Global TypedArrays to simulate C-like memory
// children: flatted array. node U's child C is at index (U * 26 + C)
const children = new Int32Array(MAX_NODES * ALPHABET);
const failLink = new Int32Array(MAX_NODES);
const dp = new Int32Array(MAX_NODES);
const isWord = new Uint8Array(MAX_NODES);

// Manual Queue for BFS
const q = new Int32Array(MAX_NODES);

let nodesCount = 1; // Start with 1 (Root is 0)

// -------------------------------------------------------------------------
// HELPER FUNCTIONS
// -------------------------------------------------------------------------

// Efficiently clear a node's data for reuse
// We assume 'node' index is being allocated for the first time in this test case
function clearNode(node) {
  const start = node * ALPHABET;
  // We must clear the children slots because they might contain
  // "DFA shortcuts" from the previous test case.
  // Hand-unrolling or fill is safer here.
  children.fill(0, start, start + ALPHABET);

  failLink[node] = 0;
  dp[node] = 0;
  isWord[node] = 0;
}

// -------------------------------------------------------------------------
// MAIN LOGIC
// -------------------------------------------------------------------------

// Read all stdin into a buffer
const buffer = fs.readFileSync(0);
let bufferIdx = 0;
const bufLen = buffer.length;

// Helper to skip whitespace and return if EOF
function skipWhitespace() {
  while (bufferIdx < bufLen && buffer[bufferIdx] <= 32) {
    bufferIdx++;
  }
  return bufferIdx < bufLen;
}

// Read integer
function readInt() {
  if (!skipWhitespace()) return null;

  let res = 0;
  let charCode = buffer[bufferIdx];
  while (bufferIdx < bufLen && charCode >= 48 && charCode <= 57) {
    res = res * 10 + (charCode - 48);
    bufferIdx++;
    if (bufferIdx < bufLen) charCode = buffer[bufferIdx];
  }
  return res;
}

// Consume a word from buffer and insert directly into Trie
// Returns nothing, side-effect only
function readAndInsertWord() {
  if (!skipWhitespace()) return;

  let curr = 0; // Root
  let charCode = buffer[bufferIdx];

  // Read 'a'-'z' (97-122)
  while (bufferIdx < bufLen && charCode > 32) {
    const idx = charCode - 97; // 'a' -> 0

    // Calculate flat index for children[curr][idx]
    const flatIdx = curr * ALPHABET + idx;

    if (children[flatIdx] === 0) {
      // Allocate new node
      clearNode(nodesCount);
      children[flatIdx] = nodesCount;
      nodesCount++;
    }

    curr = children[flatIdx];

    bufferIdx++;
    if (bufferIdx < bufLen) charCode = buffer[bufferIdx];
  }

  isWord[curr] = 1;
}

function solve() {
  let head = 0;
  let tail = 0;
  let maxChain = 0;

  // 1. Initialize Root's children (Depth 1)
  // Root is 0. failLink[0] is 0. dp[0] is 0.

  // Iterate root's children manually
  // Root's children start at index 0 in the flat array
  for (let i = 0; i < ALPHABET; i++) {
    const child = children[i]; // children[0 * 26 + i]
    if (child !== 0) {
      failLink[child] = 0;
      if (isWord[child] === 1) {
        dp[child] = 1;
        if (maxChain < 1) maxChain = 1;
      } else {
        dp[child] = 0;
      }
      q[tail++] = child;
    }
  }

  // 2. BFS
  while (head < tail) {
    const u = q[head++];
    const uFail = failLink[u];
    const uDp = dp[u];

    const uBase = u * ALPHABET;
    const failBase = uFail * ALPHABET;

    for (let i = 0; i < ALPHABET; i++) {
      const v = children[uBase + i];
      const failTarget = children[failBase + i]; // Equivalent to trie[fail[u]][i]

      if (v !== 0) {
        // Child exists (Real Edge)
        failLink[v] = failTarget;

        // DP Logic
        let val = uDp;
        const suffixVal = dp[failTarget];
        if (suffixVal > val) val = suffixVal;

        if (isWord[v] === 1) val++;

        dp[v] = val;
        if (val > maxChain) maxChain = val;

        q[tail++] = v;
      } else {
        // Child doesn't exist: DFA shortcut (Flattening)
        // Point directly to the failure's child
        children[uBase + i] = failTarget;
      }
    }
  }
  return maxChain;
}

// -------------------------------------------------------------------------
// EXECUTION
// -------------------------------------------------------------------------

while (true) {
  const n = readInt();
  if (n === null || n === 0) break;

  // Reset global cursor
  nodesCount = 1;
  clearNode(0); // Clear root

  // Read N words
  for (let i = 0; i < n; i++) {
    readAndInsertWord();
  }

  console.log(solve());
}
